driver/instantiator.ml

```diff
@@ -0,0 +1,184 @@
+module CU = Compilation_unit
+
+type error =
+  | Not_compiled_as_argument of CU.t
+  | Incorrect_target_filename of {
+      expected_basename : Misc.filepath;
+      expected_extension : string;
+      actual_basename : Misc.filepath;
+      compilation_unit : CU.t;
+    }
+  | Not_parameterised of CU.t
+  | Missing_argument of { param : Global_module.Name.t }
+  | No_such_parameter of {
+      base_unit : CU.t;
+      param : Global_module.Name.t;
+      arg : Global_module.Name.t
+    }
+  | Repeated_parameter of {
+      param : Global_module.Name.t;
+      arg1 : CU.t;
+      arg2 : CU.t;
+    }
+
+
+exception Error of error
+
+let error e = raise (Error e)
+
+type unit_info = {
+  ui_unit : CU.t;
+  ui_arg_descr : Lambda.arg_descr option;
+  ui_format : Lambda.module_block_format;
+}
+
+let instantiate
+      ~src ~args targetcm ~expected_extension ~read_unit_info ~compile =
+  let unit_infos = read_unit_info src in
+  let base_compilation_unit = unit_infos.ui_unit in
+  let arg_info_of_cm_path cm_path =
+    let unit_infos = read_unit_info cm_path in
+    match unit_infos.ui_arg_descr with
+    | None -> error (Not_compiled_as_argument unit_infos.ui_unit)
+    | Some { arg_param; arg_block_field } ->
+      arg_param, (unit_infos.ui_unit, arg_block_field)
+  in
+  let arg_info = List.map arg_info_of_cm_path args in
+  let arg_pairs : CU.argument list =
+    List.map
+      (fun (param, (value, _)) : CU.argument ->
+         { param = CU.of_global_name param; value })
+      arg_info
+  in
+  let arg_map : (CU.t * int) Global_module.Name.Map.t =
+    match Global_module.Name.Map.of_list_checked arg_info with
+    | Ok map -> map
+    | Error (Duplicate { key; value1 = (arg1, _); value2 = (arg2, _) }) ->
+      error (Repeated_parameter { param = key; arg1; arg2 })
+  in
+  let compilation_unit = CU.create_instance base_compilation_unit arg_pairs in
+  let expected_output_prefix = CU.base_filename compilation_unit in
+  let output_prefix = Filename.remove_extension targetcm in
+  let output_prefix_basename = Filename.basename output_prefix in
+  if not (String.equal output_prefix_basename expected_output_prefix)
+  then begin
+    (* The module will only work if given the correct filename *)
+    error (Incorrect_target_filename
```

Comment:
I'm unfamiliar with how compilation units usually works. In vanilla ocaml, I can compile `a.ml` into `b.cmi/cmo`? Does that interfere with future code refer to `module A`?

Reply:
Yes, you can, and it does: other modules have to say `B`. (We actually use this all the time to compile libraries.) We don't have that degree of freedom with `-instantiate` since .cmx lookup has to work.